R/input-switch.R

```diff
@@ -78,16 +78,48 @@ toggle_switch <- function(id, value = NULL, session = get_current_session()) {
 }
 
 # Keep this internal for now until we agree on the UX
-input_dark_mode_switch <- function(id, label, ..., width = "auto") {
-  res <- input_switch(id, label, ..., width = width)
-  # TODO: .navbar should probably do this for .shiny-input-container
-  res <- tagAppendAttributes(res, style = "margin-bottom: 0;")
-  tagAppendAttributes(
-    res, .cssSelector = "input",
-    onclick = "document.documentElement.setAttribute('data-bs-theme', this.checked ? 'dark' : 'light'); $(window).resize();"
+input_dark_mode_switch <- function(id, ..., mode = NULL) {
```

Comment:
I think that instead of `mode`, this should be `value`, and the HTML attribute should also be named `value`. (Or, alternatively, the HTML attribute can be just called `mode`.)

I think that, as much as is practical, we should aim to have the R (and Python) functions to be thin wrappers for HTML.

Reply:
One more thing: I think `id` should not be required. For most use cases, the app author will just want to put the switch on the page and won't care about reading the value. For those cases, there will be one less thing for the author to think about.

Reply:
➕ ➕ I just came to the same conclusion re `value` and I agree that `id` should be optional.

Reply:
Actually, after thinking about this a bit more I'm not sure I agree with the name change to `value` on the R side. In general, I agree with the principle of keeping the R/Python APIs close to the HTML markup.

But in this case, `value` isn't as meaningful as mode. Especially since it's in the function name name – i.e. dark _mode_. Also the word "theme" is quite saturated in bslib/Bootstrap/Bootswatch, which necessitated us finding `preset` as a new word for _collection of settings for Bootstrap_. Also, Bootstraps documentation refer to _color modes_ and there is a reasonable likelihood we'll end up with an `input_color_mode()` (or a component that can toggle between more than just dark/light).

So from the user's perspective, I think `mode` is the right word that balances everything out. I'll nudge the component logic in this direction, too.

Reply:
The idea of making `id` optional is interesting, but I'm worried it's going to have cascading effects that will significantly alter the implementation and raise some knock-on issues.

In particular, if we were to also watch the DOM element with the reflected preference across toggle buttons, we might not even need an `id` at all. Since they'd all share the same state and changing one of them would change any of them, maybe they should all have the same `id` (or an `id` that's completely hidden from the user). In that case, maybe we shouldn't have an `input` at all, but we could pass the color mode or system preference back via `session$clientData`. And in that case, we wouldn't need to use an input binding at all for this component, it could instead register a custom message handler.

A reason for making `id` explicit is that we may want to scope the color mode to regions of the DOM. Currently, we're setting this value on the `documentElement`, but it could just as easily work on any subtree of the DOM. I considered adding this – the component would take a selector to determine which attribute is updated – because it's [part of Bootstrap's color mode design](https://getbootstrap.com/docs/5.3/customize/color-modes/#example). In this case, having an `id` would be important to be able to differentiate between the global theme and the localized theme.

Cutting between all of these options: I think we should move `id` after the `...` so that it's required to be named. That makes it optional, while giving us the ability to deprecate or adjust how its used in the future.

Reply:
I made `id` optional here: d2a56f289d4f39c9a24a6c13dcf4c8146f89b3b2. For now I've said in the docs that it's optional but required for server-side interactions.

Reply:
c181ea50b2fa8386bdec6c047ce8300724e34bfa takes this one step further and removes the need for `id` from `toggle_dark_mode()` as well, since all dark mode inputs observe the document element attribute.